app/modules/pwm2.c

```diff
@@ -0,0 +1,553 @@
+// Module for interfacing with PWM
+
+#include <stdlib.h>
+#include "module.h"
+#include "lauxlib.h"
+#include "platform.h"
+#include "c_types.h"
+#include "pin_map.h"
+#include "mem.h"
+#include "hw_timer.h"
+
+#define PWM2_TMR_OWNER ((os_param_t)'2')
+#define PWM2_TMR_MAGIC_80MHZ 16
+#define PWM2_TMR_MAGIC_160MHZ 32
+
+//############################
+// global type definitions
+
+typedef struct
+{
+  uint32_t offInterruptCounter;
+  uint32_t pulseInterruptCcounter;
+  uint32_t currentInterruptCounter;
+  uint16_t gpioMask;
+} pwm2_pin_interrupt_t;
+
+typedef struct
+{
+  pwm2_pin_interrupt_t pin[GPIO_PIN_NUM];
+  uint16_t enabledGpioMask;
+} pwm2_interrupt_handler_data_t;
+
+typedef struct
+{
+  uint32_t pulseResolutions;
+  uint32_t divisableFrequency;
+  uint32_t frequencyDivisor;
+  uint32_t duty;
+  uint32_t resolutionCPUTicks;
+  uint32_t resolutionInterruptCounterMultiplier;
+} pwm2_pin_setup_t;
+
+typedef struct
+{
+  pwm2_pin_setup_t pin[GPIO_PIN_NUM];
+  uint32_t interruptTimerCPUTicks;
+  uint32_t interruptTimerTicks;
+  bool isStarted;
+} pwm2_setup_data_t;
+
+typedef struct
+{
+  pwm2_interrupt_handler_data_t interruptData;
+  pwm2_setup_data_t setupData;
+} pwm2_module_data_t;
+
+//############################
+// module vars, lazy initialized, allocated only if pwm2 is being used
+
+static pwm2_module_data_t *moduleData = NULL;
+
+//############################
+
+static bool isPinSetup(const pwm2_module_data_t *data, const uint8_t pin)
+{
+  return data->setupData.pin[pin].pulseResolutions > 0;
+}
+
+static uint32_t getCPUTicksPerSec()
+{
+  return system_get_cpu_freq() * 1000000;
+}
+
+static uint8_t getCpuTimerTicksDivisor()
+{
+  return system_get_cpu_freq() == 80 ? PWM2_TMR_MAGIC_80MHZ : PWM2_TMR_MAGIC_160MHZ;
+}
+
+static uint32_t findGCD(uint32_t n1, uint32_t n2)
```

Comment:
Yuk.  even Euclid uses remainder here.  Time for a couple of % operators. :smile:

Reply:
he is being giant even millenniums old. i yet have to reach even half of century ;) anyhow, would now even euclidus be satisfied with this version?